### ISSUE TITLE
Block Pattern loads super slow in the case of Pattern created with ACF

### ISSUE THREAD

**Issue by CreativeDive** [AdvancedCustomFields/acf#583]
Hey @lgladdy @mattgrshaw,

I have created a collection of useful block pattern. I love pattern by the way. But all pattern created with ACF will loaded super slow if I click on the pattern selection.

Here you can see, the first two pattern are created with core blocks. This first two are there immedially. All other pattern below created with ACF will load toooo slow. More than 20 seconds to load 5 additional ACF block pattern.

![Nov-23-2021 21-39-58](https://user-images.githubusercontent.com/31563167/143107282-f4f87526-db92-4228-b2a3-d97305151f75.gif)

This is very annoying and a bad user experience to use pattern in this case.

Is there any solution to use something like cached block pattern if they are created with ACF or other methods to get this more faster please?

**Comment by lgladdy** [comment]
Hey @CreativeDive

Yeh, we're aware of this issue; it's certainly not an optimum experience at the moment! This is also an issue for things like block styles where React tries to render live.

We're preparing for a suite of block improvements and fixes for WordPress 5.9, so we'll see how viable getting some kind of temporary cache of blocks in place for that release!

**Comment by CreativeDive** [comment]
@lgladdy thank you for response. Yes I think cache can be powerful here. But if caching is a solution here, we have to need a solution to refresh this cache.

The block pattern are rendered live, with global defined styles like color, font sizes ...

If there is a cache, the global defined styles could be displayed wrong, if the user has changed global styles.

For this case, something like "acf_clear_block_pattern_cache()" function  would be great here. So developers can call this function if global styles has been changed.

**Comment by lgladdy** [comment]
@CreativeDive I was actually thinking about this issue last night while trying to get to sleep! I was wondering if we could do something based on a hash of the template to build the cache, so if the template changes, we know to re-render. That wouldn't necessarily help with CSS changes of course - so you're right we'd need a way to clear that cache.

ACF doesn't really have any kind of admin screen for ACF operations yet, so we'll need to think a bit about how that would work, as we can't rely on users to fire a function to clear caches - a lot of people use blocks because they're not developers who want to worry about PHP.

**Comment by CreativeDive** [comment]
@lgladdy yes I can understand. A possibility would be to refresh the cache automatically each time after an ACF option page has been saved. I know this is not a solution in all cases. But for other cases (developer case) an additional function to do that would be very helpful.

**Comment by lgladdy** [comment]
@CreativeDive I think it might well be enough to store the preview cache inside the block editor instance itself. So first load of a block preview would be cached, then every other load of that would use the local cache, given block patterns are likely to contain repeated blocks.

We'll do some testing anyway - hopefully we'll have a solution for our next beta before 5.9 so you can test!

**Comment by CreativeDive** [comment]
@lgladdy thank you. I'm sure you will find a good solution. I'm here for testing if you have a new beta ready.

**Comment by lgladdy** [comment]
@CreativeDive Can I check what browser you're using please? Performance seems to vary significantly between the ones I've tested - while we need to fix the AJAX call universally, I'm wondering if there's another issue here too!

**Comment by CreativeDive** [comment]
@lgladdy I think the browser is not the problem here. I have the same performance issues with Google Chrome or Safari. The important thing how fast a bundle of ACF block pattern can be loaded is the performance of the hosting provider. In my case I don't not working with power hosting like Amazon AWS. I could see cases on power hosted websites and the pattern are loaded faster.

But power hosting is not the reality. A huge amount of websites are hosted on shared server with limited resources. And here the pattern are loading super slow if we have a bundle more than 20 ACF pattern. But in real life, the amount of available pattern will increase in the future. Therefore a performance boost is very important here.

**Comment by lgladdy** [comment]
I'm asking what browser you're using to ensure we solve the issue and test it in your browser too.

The performance in Chrome seems significantly better than Safari, for example. I don't want to commit a fix that I believe solves it only to find there is another issue that's specific to a browser.

**Comment by CreativeDive** [comment]
@lgladdy Currently I‘m away from the Office Keyboard. But I will share the information about the browser later this day.

**Comment by nick6352683** [comment]
I had similar experience a year or so ago. The ACF patterns were getting loaded in about 30 seconds. The reason was the Editor Plus plugin, which now has been sold to the people who also bought the EditorsKit plugin, and in both cases added some completely useless stuff and destroyed both plugins, as in many case they are not compatible with other plugins such as Getwid.


Make you you test your patterns without any other "Gutenberg" plugins being active as a starting point of debugging...

**Comment by CreativeDive** [comment]
@lgladdy I've tested it again on both browsers, it's nearly the same in my case. I have up to 20 patterns, 18 patterns build with ACF and the loading time is 47 seconds for all in both browsers.

Google Chrome - Version 97.0.4692.99
Safari - Version 15.1 (17612.2.9.1.20)

You can see in the screen recording below it takes a lot of time.

https://user-images.githubusercontent.com/31563167/151229797-a59deccf-406c-4603-b81d-77171c9f1487.mp4

@nick6352683 I don't know if we have the same case and the same type of patterns? My patterns includes a bundle of nested blocks and each block will loaded via AJAX. How many AJAX request can be loaded in the same time? I think around 5 requests. I have no Gutenberg plugin active and no other block plugins. My installation is working with WP 5.9 and the latest ACF beta. Do you have no slow loading times by loading patterns like me?

**Comment by nick6352683** [comment]
@CreativeDive ,

I am sorry, this is embarrassing. I had misunderstood your issue...

My issue was inserting a pattern in the editor, not the speed of the previews in the Pattern List. For me, your type of issue was a big issue with WP 5.9 betas, and RCs... with Regular ACF blocks, as I have 5 Categories and a total of 51 ACF Blocks. Things got dramatically better with WP 5.9 released yesterday. I wonder of you have the Gutenberg plugin active or not (I don't, as the latest Gut. plugin introduces some bugs that I have reported to the devs). My only advice right now is to deactivate the Gut. plugin if active and see if things are better. Block listing speed was never an issue for me before 5.9... As far as Patterns go, I only have 2 Categories with only 2 patterns in each category, just for testing things out. I'm waiting for Gutenberg to stabilize (only after 3 years!), as regularly I have to update the patterns as Gutenberg keeps breaking my layouts every few months... I've been told that WP 5.9.1 will be out shortly with many bugs to be fixed, introduced by WP5.9.

For testing, I'll quickly create 10 or so Patterns that include ACF blocks and I'll report back here tomorrow... and yes, my patterns have many nested blocks and complex layouts.

I wonder if we can disable the pattern's live previews, and show static images instead, that's how I preview my ASF blocks.

**Comment by nick6352683** [comment]
@CreativeDive ,

After some further testing, I found a way to avoid this issue. This happens many times (but not always) when trying to browse the Blocks/Patterns from the main inserter. Again, this does not always happen, but most of the times. 

In order to avoid this performance issue, click on any inserter in the editor, and then click on the "Browse All" button from the popup. The main inserter will open and all the blocks/patterns will be displayed. Here is a 3 minute video demo., the last Group of blocks should be the Category "Woocommerce"...

https://www.youtube.com/watch?v=miHvOza6jzA

This is a Gutenberg issue, and I don't think this is an ACF issue in my opinion...

**Comment by CreativeDive** [comment]
@nick6352683 thanks for the input. But the topic of this issue is the loading time of the block "PATTERN", not of the blocks. :-)

**Comment by CreativeDive** [comment]
@nick6352683 I think I can understand your statement. If you have loaded the patterns completely the first time, the second  time all patterns are loaded faster. BUT, if I reload the page or add a new page / post. The same again, patterns are loaded again and it takes a lot of time, each time.

My Goal is to make a set about of 50-100 usable patterns. But currently each time the user is waiting some minutes to load 50-100 patterns. This should be faster, definitely.

**Comment by CreativeDive** [comment]
@nick6352683 

To compare both, here another test:

Loading 27 default pattern builded with core blocks only ---> Loading time 15 Seconds

https://user-images.githubusercontent.com/31563167/151756299-9131c696-f32c-4926-beb7-388b4cd3882b.mp4

Loading 18 pattern builded with ACF blocks only ---> Loading time 47 Seconds

There are essential differences.

**Comment by lgladdy** [comment]
@CreativeDive Do you know how many different ACF blocks you're using inside those 18 patterns?

**Comment by CreativeDive** [comment]
@lgladdy 

This are the ACF block which are used by the patterns, around 7 block types:

Line block
Custom Teaser block
Image / Link block
Content box block
Image / Text block
Background Box block
Media Box block

For headings, text and buttons I use the core blocks:

Heading block
Paragraph block
Button block

Could you made any progress or found ways to fix this issue?

**Comment by CreativeDive** [comment]
@lgladdy 

Here another test. This are only pricing table patterns. This pattern contains only a very simple "Price block" to format the price including the currency builded with ACF blocks and a "Content box block" which includes an inner block, builded with ACF blocks. The rest of this pattern is builded with core blocks. 

https://user-images.githubusercontent.com/31563167/151763084-92660f6d-1daf-43fd-a111-5d776f532a6b.mp4

**Comment by lgladdy** [comment]
@CreativeDive That last video makes me more concerned that this isn't going to be something we can fix quickly and is a limitation of ACF Blocks. We can make all the optimisations to make sure we only load a block once, but that's only viable if absolutely nothing about the block changes. It's hard to tell from your video because it's compressed, but it looks like the top row of those blocks, for example, contain 3 of the same block. I'd hope we'll be able to get each of those loaded in one AJAX call - but there rest seem to load different results, so each of those will still need one call each.

We can also look at batching up AJAX requests if they're requested quickly by Gutenberg, but we'll need to do some testing on likely that is to improve things, as it might make the issue worse.

I don't think it would be viable to "preload" the views of those blocks either - there could be hundreds, and it would likely swap the performance issue from display to page load time - and increase, significantly, the size of the DOM which would cause more issues for users on nginx with gzip encoding.

It might be worth considering optimisations you can make on the server to help here too. Things like enabling HTTP/2, and seeing how long each call to our fetch-block AJAX endpoint is taking might give you insights into things you can improve server side while we continue to improve performance ACF's side too.

**Comment by CreativeDive** [comment]
@lgladdy 

here a very simple test. I have created a new pattern. This pattern contains a very very simple ACF block. I have created 10 pattern with the same content each time:

https://user-images.githubusercontent.com/31563167/151766178-81602a06-027b-40bc-97d2-cee0161dbf03.mp4

12 seconds loading time for a very very simple pattern with each time the same block.

That shouldn't be the end result.

**Comment by lgladdy** [comment]
@CreativeDive Yup - as I said, that's the case we can fix easily... we should be able to reduce that to one AJAX call... it's the more complicated patterns like your previous video which are still going to require a handful of calls which we can't do much about easily!

**Comment by CreativeDive** [comment]
@lgladdy maybe to reduce that to only one AJAX call could be very helpful here.

I know and the way how the pattern are loaded is very performance-heavy.

--> Iframe
--> Live rendered

Somehow these pattern have to be not only useful for the user, but also usable.

Majority of users don't have power hosting. So it cannot be assumed that users can optimize anything in terms of server performance. In the end, the user always complains about the long loading times of the pattern.

**Comment by lgladdy** [comment]
@CreativeDive Yeh, I totally understand. And the improvements to caching we're planning for 5.12 will make your example significantly faster - it might just not be as fast as we'd like!

There's a couple of other options I can think of, things like silently preloading block pattern blocks after the page loads in the background, but that's the kind of fundamental refactor of some of the code that might not make it for 5.12.

**Comment by CreativeDive** [comment]
@lgladdy Any kind of improvement is very welcome, even if it is not optimal at first. Thanks for your time.

Preloading in the background sounds really cool, but definitely this is a fundamental refactor, not easy and not fast actionable.
